staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/grpc_service_unix_test.go

```diff
@@ -36,17 +38,139 @@ const (
 	endpoint = "unix:///@kms-socket.sock"
 )
 
+// TestKMSPluginLateStart tests the scenario where kms-plugin pod/container starts after kube-apiserver pod/container.
+// Since the Dial to kms-plugin is non-blocking we expect the construction of gRPC service to succeed even when
+// kms-plugin is not yet up - dialing happens in the background.
+func TestKMSPluginLateStart(t *testing.T) {
+	callTimeout := 3 * time.Second
+
+	service, err := NewGRPCService(endpoint, callTimeout)
+	if err != nil {
+		t.Fatalf("failed to create envelope service, error: %v", err)
+	}
+	defer destroyService(service)
+
+	time.Sleep(callTimeout / 2)
+	f, err := startFakeKMSProvider(kmsapiVersion)
+	if err != nil {
+		t.Fatalf("failed to start test KMS provider server, error: %v", err)
+	}
+	defer f.server.Stop()
+
+	data := []byte("test data")
+	_, err = service.Encrypt(data)
+	if err != nil {
+		t.Fatalf("failed when execute encrypt, error: %v", err)
+	}
+}
+
+// TestIntermittentConnectionLoss tests the scenario where the connection with kms-plugin is intermittently lost.
+func TestIntermittentConnectionLoss(t *testing.T) {
+	var (
+		wg1      sync.WaitGroup
+		wg2      sync.WaitGroup
+		timeout  = 30 * time.Second
+		blackOut = 1 * time.Second
+		data     = []byte("test data")
+	)
+	// Start KMS Plugin
+	f, err := startFakeKMSProvider(kmsapiVersion)
+	if err != nil {
+		t.Fatalf("failed to start test KMS provider server, error: %v", err)
+	}
+
+	//  connect to kms plugin
+	service, err := NewGRPCService(endpoint, timeout)
+	if err != nil {
+		t.Fatalf("failed to create envelope service, error: %v", err)
+	}
+	defer destroyService(service)
+
+	_, err = service.Encrypt(data)
+	if err != nil {
+		t.Fatalf("failed when execute encrypt, error: %v", err)
+	}
+	t.Log("Connected to KMSPlugin")
+
+	// Stop KMS Plugin - simulating connection loss
+	f.server.Stop()
+	t.Log("KMS Plugin is stopped")
+
+	wg1.Add(1)
+	wg2.Add(1)
+	go func() {
+		defer wg2.Done()
+		// Call service to encrypt data.
+		t.Log("Sending encrypt request")
+		wg1.Done()
+		_, err := service.Encrypt(data)
+		if err != nil {
+			t.Fatalf("failed when executing encrypt, error: %v", err)
+		}
+	}()
+
+	wg1.Wait()
+	time.Sleep(blackOut)
+	// Start KMS Plugin
+	f, err = startFakeKMSProvider(kmsapiVersion)
+	if err != nil {
+		t.Fatalf("failed to start test KMS provider server, error: %v", err)
+	}
+	defer f.server.Stop()
+	t.Log("Restarted KMS Plugin")
+
+	wg2.Wait()
+}
+
+func TestUnsupportedVersion(t *testing.T) {
```

Comment:
Somewhere in one of these tests can we please exercise a large number of concurrent requests, to make sure the locking has some coverage in unit tests?

Reply:
It is unlikely that I could do a better job than GO's race detector. All tests are passing with race flag on (I believe that race detection is on by default in K8S CI).
What do you think?